trieste/acquisition/function.py

```diff
@@ -201,6 +203,79 @@ def update_acquisition_function(
         return function
 
 
+class DeepGaussianProcessExpectedImprovement(SingleModelAcquisitionBuilder):
+    """
+    Builder for the expected improvement function for use with a
+    :class:`~trieste.models.gpflux.DeepGaussianProcess` model, where we have to obtain a Monte Carlo
+    estimate of the expected improvement. The "best" value is taken to be the minimum of the
+    posterior mean at observed points.
+    """
+
+    def __init__(self, sample_size: int):
+        """
+        :param sample_size: The number of samples for each batch of points.
+        :raise tf.errors.InvalidArgumentError: If ``sample_size`` is not positive.
+        """
+        tf.debugging.assert_positive(sample_size)
+
+        super().__init__()
+
+        self._sample_size = sample_size
+
+    def __repr__(self) -> str:
+        """"""
+        return f"DeepGaussianProcessExpectedImprovement({self._sample_size!r})"
+
+    def prepare_acquisition_function(
+        self, dataset: Dataset, model: ProbabilisticModel
+    ) -> AcquisitionFunction:
+        """
+        :param dataset: The data from the observer. Must be populated.
+        :param model: The model over the specified ``dataset``. Must have event shape [1] and must
+            be an instance of :class:`~trieste.models.gpflux.DeepGaussianProcess`.
+        :return: The estimated *expected improvement* acquisition function.
+        :raise ValueError (or InvalidArgumentError): If ``dataset`` is not populated, ``model``
+            does not have an event shape of [1] or is not an instance of
+            :class:`~trieste.models.gpflux.DeepGaussianProcess`.
+        """
+        if not isinstance(model, DeepGaussianProcess):
+            raise ValueError("Model must be a trieste.models.gpflux.DeepGaussianProcess.")
+
+        tf.debugging.assert_positive(len(dataset))
+
+        at = dataset.query_points
+        for i, layer in enumerate(model.model_gpflux.f_layers[:-1]):
+            mean, var = layer.predict(at, full_cov=False, full_output_cov=False)
+
+            at = mean + tf.sqrt(var) * tf.random.normal(
+                [self._sample_size, 1, tf.shape(mean)[-1]], dtype=tf.float64
+            )
+
+        mean_samples, _ = model.model_gpflux.f_layers[-1].predict(
+            at, full_cov=False, full_output_cov=False
+        )
+
+        mean = tf.reduce_mean(mean_samples, axis=0)
+
+        tf.debugging.assert_shapes(
+            [(mean, ["_", 1])], message="Expected model with event shape [1]."
+        )
+
+        eta = tf.reduce_min(mean, axis=0)
+        sampler = DeepGaussianProcessSampler(self._sample_size, model)
```

Comment:
following the more general approach, this could be an input to the class instead of being defined here